src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py

```diff
@@ -0,0 +1,783 @@
+"""
+
+GPT_NEOX_INPUTS_DOCSTRING = r"""
+    Args:
+        input_ids (`numpy.ndarray` of shape `(batch_size, input_ids_length)`):
+            `input_ids_length` = `sequence_length`. Indices of input sequence tokens in the vocabulary.
+
+            Indices can be obtained using [`AutoTokenizer`]. See [`PreTrainedTokenizer.encode`] and
+            [`PreTrainedTokenizer.__call__`] for details.
+
+            [What are input IDs?](../glossary#input-ids)
+        attention_mask (`numpy.ndarray` of shape `(batch_size, sequence_length)`, *optional*):
+            Mask to avoid performing attention on padding token indices. Mask values selected in `[0, 1]`:
+
+            - 1 for tokens that are **not masked**,
+            - 0 for tokens that are **masked**.
+
+            [What are attention masks?](../glossary#attention-mask)
+        position_ids (`numpy.ndarray` of shape `(batch_size, sequence_length)`, *optional*):
+            Indices of positions of each input sequence tokens in the position embeddings. Selected in the range `[0,
+            config.max_position_embeddings - 1]`.
+        past_key_values (`Dict[str, np.ndarray]`, *optional*, returned by `init_cache` or when passing previous `past_key_values`):
+            Dictionary of pre-computed hidden-states (key and values in the attention blocks) that can be used for fast
+            auto-regressive decoding. Pre-computed key and value hidden-states are of shape *[batch_size, max_length]*.
+        output_attentions (`bool`, *optional*):
+            Whether or not to return the attentions tensors of all attention layers. See `attentions` under returned
+            tensors for more detail.
+        output_hidden_states (`bool`, *optional*):
+            Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors for
+            more detail.
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+"""
+
+
+def rotate_half(hidden_states):
+    first_half = hidden_states[..., : hidden_states.shape[-1] // 2]
+    second_half = hidden_states[..., hidden_states.shape[-1] // 2 :]
+    return jnp.concatenate((-second_half, first_half), axis=-1)
+
+
+class FlaxGPTNeoXRotaryEmbedding(nn.Module):
+    dim: int
+    max_position_embeddings: int
+    base: int = 10000
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.inv_freq = 1.0 / (self.base ** (jnp.arange(0, self.dim, 2).astype(self.dtype) / self.dim))
+        self.cos_cached, self.sin_cached = self._compute_cos_sin(self.max_position_embeddings)
+
+    def _get_cos_sin_cache(self, seq_len):
+        if seq_len > self.max_position_embeddings:
+            return self._compute_cos_sin(seq_len)
+        else:
+            return self.cos_cached, self.sin_cached
+
+    def _compute_cos_sin(self, seq_len):
+        t = jnp.arange(seq_len, dtype=self.inv_freq.dtype)
```

Comment:
Note to reviewer: single-letter variables chosen to maintain equivalence with the PyTorch modelling code https://github.com/huggingface/transformers/blob/2c658b5a4282f2e824b4e23dc3bcda7ef27d5827/src/transformers/models/gpt_neox/modeling_gpt_neox.py#L303